### ISSUE TITLE
format_ad_timedelta throws exception

### ISSUE THREAD

**Issue by dirkjanm** [cannatag/ldap3#639]
v2.5.2 seems to break on both python 2 and 3 with the new formatter for AD timedelta objects:

```
File "/home/dirkjan/.local/share/virtualenvs/venv/local/lib/python2.7/site-packages/ldap3/operation/search.py", line 453, in checked_attributes_to_dict_fast
    checked_attributes[name] = format_attribute_values(schema, name, decode_raw_vals_fast(attribute[3][1][3]) or [], custom_formatter)
  File "/home/dirkjan/.local/share/virtualenvs/venv/local/lib/python2.7/site-packages/ldap3/protocol/formatters/standard.py", line 213, in format_attribute_values
    formatted_values = [formatter(raw_value) for raw_value in values]  # executes formatter
  File "/home/dirkjan/.local/share/virtualenvs/venv/local/lib/python2.7/site-packages/ldap3/protocol/formatters/formatters.py", line 338, in format_ad_timedelta
    return format_ad_timestamp(raw_value * -1) - format_ad_timestamp(0)
TypeError: unsupported operand type(s) for -: 'str' and 'datetime.datetime'
```

In the call stack `raw_value` is a bytes object, which is multiplied with -1 (an integer). This will not produce a valid result and thus the `format_ad_timestamp` function will error out and return a bytes object, from which a `datetime.datetime` object is then substracted, which raises this exception.

In a related note, adding this filter changes the output for some objects from an int object to a datetime.datetime object (which is a bit unexpected for a patch release), which breaks tools that use the library and expect an int object.

**Comment by cannatag** [comment]
HI, this code was submitted by a user, I cannot check it with AD. Do you mean that it always fails?

**Comment by dirkjanm** [comment]
Yes it always fails. Seems like the user who submitted it did not test it properly.

**Comment by yaap7** [comment]
Same issue here using v2.5.2 with python3.

```
[long error stack]
  File "/home/yaap/.local/lib/python3.5/site-packages/ldap3/protocol/formatters/standard.py", line 213, in format_attribute_values
    formatted_values = [formatter(raw_value) for raw_value in values]  # executes formatter
  File "/home/yaap/.local/lib/python3.5/site-packages/ldap3/protocol/formatters/standard.py", line 213, in <listcomp>
    formatted_values = [formatter(raw_value) for raw_value in values]  # executes formatter
  File "/home/yaap/.local/lib/python3.5/site-packages/ldap3/protocol/formatters/formatters.py", line 337, in format_ad_timedelta
    return format_ad_timestamp(raw_value * -1) - format_ad_timestamp(0)
TypeError: unsupported operand type(s) for -: 'bytes' and 'datetime.datetime'
```

Funny to see I encountered this error while working on the same kind of project as dirkjanm ^^.

**Comment by may-day** [comment]
The code apparently deals with integer values while they are actually of type bytes (or str in py2). the validation code and testunit  also use integer...
so the format_ad_timedelta would work with sth like:

def format_ad_timedelta(raw_value):
    if raw_value.startswith(b'-'):
        unsigned_raw_value = raw_value[1:]
    else:
        unsigned_raw_value = raw_value
    return format_ad_timestamp(unsigned_raw_value) - format_ad_timestamp(b'0')

which looks kinda ugly :)

**Comment by djlambert** [comment]
I'm seeing this issue also. The `groupType` attribute in Active Directory is a negative number

**Comment by jynolen** [comment]
A simple 
`def format_ad_timedelta(raw_value):
    if isinstance(raw_value, bytes):
        if raw_value * -1 == b"":
              raw_value=int(raw_value)
    return format_ad_timestamp(raw_value* -1) - format_ad_timestamp(0)
`

**Comment by Selora** [comment]
Here's my take on this. Seems to resolves the issue.
`return format_ad_timestamp(0) if not int(raw_value) else format_ad_timestamp(~int(raw_value)) - format_ad_timestamp(0)`

**Comment by may-day** [comment]
hmm, function format_ad_timestamp itself compares it's parameter with a byte-string (the max-value), so maybe it is better not calling it with an integer.

**Comment by cannatag** [comment]
Any idea on this? I cannot test it because I don't have access to AD. Could anyone extend the relevant  tests checking the correct behaviour? With a valid test suite I can go on on debugging. 

Thanks,
Giovanni

**Comment by yaap7** [comment]
I tried to work on this one but the version available from pip3 is version 2.5.2, which contains the issue, whereas the version I got from a `git clone` does not contain the function `format_ad_timedelta` in `./ldap3/protocol/formatters/formatters.py`.

How could I help on this case?
Is it already fixed in the git version?

Thanks,
yaap

**Comment by cannatag** [comment]
I removed it, you can add a correct version in the same file.

This is the original function:
```
def format_ad_timedelta(raw_value):
    """
    Convert a negative filetime value to a timedelta.
    """
    # Active Directory stores attributes like "minPwdAge" as a negative
    # "filetime" timestamp, which is the number of 100-nanosecond intervals that
    # have elapsed since the 0 hour on January 1, 1601. By making the number
    # positive, we can reuse format_ad_timestamp to get a datetime object.
    # Afterwards, we can subtract a datetime representing 0 hour on January 1,
    # 1601 from the returned datetime to get the timedelta.
    return format_ad_timestamp(raw_value * -1) - format_ad_timestamp(0)
```

Thanks.

**Comment by yaap7** [comment]
I tried a fix in pull request #660 .

As @may-day said, the `format_ad_timestamp` takes a byte as argument so I prefer to correct the raw_value in `format_ad_timestamp` directly instead of `format_ad_timedelta`.
("correct the raw_value" = reverse the negative value using `raw_value * -1`)

I think it could be a correct solution so I also add the formatter to the `lockOutObservationWindow` attribute since it uses the same format as `lockOutDuration`.
Tested with python3 and python2.

Please tell me what you think about this PR.

Thanks,
yaap

**Comment by cannatag** [comment]
Fixed in next release by yaap.

**Comment by cannatag** [comment]
Fixed in version 2.6.

**Comment by bl4ckarch** [comment]
I've opened PR #1191  to fix this issue. The fix has been tested successfully with bloodhound-python against Active Directory and resolves the TypeError completely.

Would appreciate review from maintainers. Happy to make any adjustments needed.
